File: SamplingDevice/app/src/main/python/estimate_lightsource.py
import cv2 as cv
import numpy as np
from matplotlib import pyplot as plt
import math

RW_LIGHTSIZE = 330
UNITY_SCALE = 1
THRESHOLD = 230
FOCAL_LENGTH_IN_PIXELS = 2625.591533201383

def estimate(image_bytes):
    np_arr = np.frombuffer(image_bytes, dtype=np.uint8)
    img_color = cv.imdecode(np_arr, cv.IMREAD_UNCHANGED)
    img = cv.cvtColor(img_color, cv.COLOR_BGR2GRAY)
    
    if img is None:
        raise ValueError("Failed to decode image")
        
    img = cv.flip(img, 1)
    height, width = img.shape

    ## threshold image
    return_value, threshold_binary = cv.threshold(img, THRESHOLD, 255, 0)
    moment = cv.moments(threshold_binary)
    
    ## find brightest moment
    if moment['m00'] != 0:
        centroidX = int(moment['m10'] / moment['m00'])
        centroidY = int(moment['m01'] / moment['m00'])
    else:
        centroidX, centroidY = 0, 0

    centerX = width // 2
    centerY = height // 2
    
    ## Find radius of brightest moment
    contours, hierarchy = cv.findContours(threshold_binary, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
    max_radius = 0
    max_coord = None
    for contour in contours:
        coord, radius = cv.minEnclosingCircle(contour)
        if radius > max_radius:
            max_radius = radius
            max_coord = coord

    ## find distance from image center to centroid center
    du = abs(centerX - centroidX)
    dv = abs(centerY - centroidY)


    ## Calculate mean intensity value of an image's histogram
    img_color = cv.flip(img_color, 1)
    if img_color is None:
        raise FileNotFoundError(f'image not found')
    hist = cv.calcHist([img_color], [0], None, [256], [0, 256])
    # Calculate the mean intensity value
    total_pixels = img_color.shape[0] * img_color.shape[1]
    mean_intensity = np.sum(np.array([i for i in range(256)]) * hist.flatten()) / total_pixels

    ## Calculate mean RGB of mask
    mean_light_color = cv.mean(img_color, mask=threshold_binary)


    ## find distance from camera to light source center
    if max_radius == 0:
        return {
            "ls_found": 0,
            "x" : 0,
            "y" : 0,
            "z" : 0,
            "mean_intensity" : mean_intensity,
            "r" : mean_light_color[0],
            "g" : mean_light_color[1],
            "b" : mean_light_color[2],
            "a" : mean_light_color[3]
        }
        #raise Exception("No contours or any light source representations found in image")


    diameter = max_radius * 2
    mm_per_pixels = RW_LIGHTSIZE / diameter
    u = du * mm_per_pixels
    v = dv * mm_per_pixels
    #focal_length_in_pixels = width / 2 / math.tan(120 / 2)
    distance_from_camera_to_ls = FOCAL_LENGTH_IN_PIXELS * mm_per_pixels
    dcls = math.sqrt(u ** 2 + v ** 2)
    d = math.sqrt(distance_from_camera_to_ls ** 2 - dcls ** 2)
    y = round(d / 1000 * UNITY_SCALE, 6)

    ## convert pixels to mm to Unity units
    x = round(u / 1000 * UNITY_SCALE, 6)
    z = round(v / 1000 * UNITY_SCALE, 6)
    z = z * -1 if centerY < centroidY else z


    return {
        "x" : x,
        "y" : y,
        "z" : z,
        "mean_intensity" : mean_intensity,
        "r" : mean_light_color[0],
        "g" : mean_light_color[1],
        "b" : mean_light_color[2],
        "a" : mean_light_color[3]
    }
    

File: SamplingDevice/app/src/main/java/com/uwb/simplecamera/EstimateLightSource.kt
package com.uwb.simplecamera

import android.graphics.Bitmap
import androidx.camera.core.ImageProxy
import org.opencv.android.Utils
import org.opencv.core.*
import org.opencv.imgproc.Imgproc
import kotlin.math.abs
import kotlin.math.sqrt

fun Double.roundTo(digits: Int) = "%.${digits}f".format(this).toDouble()

fun estimateImageFromBitmap(bitmap: Bitmap): Map<String, Any?> {
    val RW_LIGHTSIZE = 330.0
    val UNITY_SCALE = 1.0
    val THRESHOLD = 230.0
    val FOCAL_LENGTH_IN_PIXELS = 2625.591533201383


    val mat = Mat()
    Utils.bitmapToMat(bitmap, mat)
    //Core.rotate(mat, mat, Core.ROTATE_90_COUNTERCLOCKWISE)
    val gray = Mat()
    Imgproc.cvtColor(mat, gray, Imgproc.COLOR_BGR2GRAY)
    Core.flip(gray, gray, +1) // Flip horizontally

    val height = gray.rows()
    val width = gray.cols()

    val thresholdBinary = Mat()
    Imgproc.threshold(gray, thresholdBinary, THRESHOLD, 255.0, Imgproc.THRESH_BINARY)

    val moments = Imgproc.moments(thresholdBinary)
    val centroidX = if (moments.m00 != 0.0) (moments.m10 / moments.m00) else 0.0
    val centroidY = if (moments.m00 != 0.0) (moments.m01 / moments.m00) else 0.0

    val centerX = width / 2.0
    val centerY = height / 2.0
    val du = abs(centerX - centroidX).toDouble()
    val dv = abs(centerY - centroidY).toDouble()

    val contours = mutableListOf<MatOfPoint>()
    Imgproc.findContours(thresholdBinary.clone(), contours, Mat(), Imgproc.RETR_TREE, Imgproc.CHAIN_APPROX_SIMPLE)

    var maxRadius = 0f
    var maxCenter = Point()

    for (contour in contours) {
        val contour2f = MatOfPoint2f(*contour.toArray())
        val center = Point()
        val radius = FloatArray(1)
        Imgproc.minEnclosingCircle(contour2f, center, radius)
        if (radius[0] > maxRadius) {
            maxRadius = radius[0]
            maxCenter = center
        }
    }

    val flippedMat = Mat()
    Core.flip(mat, flippedMat, +1)
    val meanIntensity = calculateMeanIntensity(flippedMat)
    val meanColor = Core.mean(flippedMat, thresholdBinary)

    if (maxRadius == 0f) {
        return mapOf(
            "ls_found" to 0,
            "x" to 0,
            "y" to 0,
            "z" to 0,
            "mean_intensity" to meanIntensity,
            "r" to meanColor.`val`[0],
            "g" to meanColor.`val`[1],
            "b" to meanColor.`val`[2],
            "a" to meanColor.`val`[3]
        )
    }

    val diameter = maxRadius * 2
    val mmPerPixel = RW_LIGHTSIZE / diameter
    val u = du * mmPerPixel
    val v = dv * mmPerPixel
    val distanceFromCameraToLS = FOCAL_LENGTH_IN_PIXELS * mmPerPixel
    val dcls = sqrt(u * u + v * v)
    val d = sqrt(distanceFromCameraToLS * distanceFromCameraToLS - dcls * dcls)

    val x = (u / 1000.0 * UNITY_SCALE).roundTo(6)
    val y = (d / 1000.0 * UNITY_SCALE).roundTo(6)
    var z = (v / 1000.0 * UNITY_SCALE).roundTo(6)
    if (centerY < centroidY) z *= -1


    return mapOf(
        "ls_found" to 1,
        "x" to x,
        "y" to y,
        "z" to z,
        "mean_intensity" to meanIntensity,
        "r" to meanColor.`val`[0],
        "g" to meanColor.`val`[1],
        "b" to meanColor.`val`[2],
        "a" to meanColor.`val`[3]
    )
}

fun calculateMeanIntensity(flippedMat: Mat): Double {
    val hist = Mat()
    val channels = MatOfInt(0)
    val histSize = MatOfInt(256)
    val ranges = MatOfFloat(0f, 256f)
    Imgproc.calcHist(listOf(flippedMat), channels, Mat(), hist, histSize, ranges)

    val totalPixels = flippedMat.rows() * flippedMat.cols()
    var meanIntensity = 0.0
    for (i in 0 until 256) {
        meanIntensity += i * hist.get(i, 0)[0]
    }
    meanIntensity /= totalPixels
    return meanIntensity
}